File: DataEngineering/02_Data_Modeling/projects/golang/05_denormalizer_engine/main.go
package main

import (
	"fmt"
	"log"

	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"
)

// In a real-world Go data pipeline, you often rely on the RDBMS
// or an OLAP engine to handle denormalization at scale.
// Using an ORM like GORM allows you to construct the wide (denormalized) 
// structs easily from normalized tables.

// Normalized: Dimension Product
type DimProduct struct {
	ProductID   int    `gorm:"primaryKey"`
	ProductName string `gorm:"not null"`
	Category    string
}

// Normalized: Dimension Store
type DimStore struct {
	StoreID   int    `gorm:"primaryKey"`
	StoreName string `gorm:"not null"`
	City      string
}

// Normalized: Fact Sales
type FactSales struct {
	SaleID    int `gorm:"primaryKey"`
	ProductID int
	StoreID   int
	Amount    float64
}

// Denormalized Wide Struct (Target for Analytics/ML/API)
type DenormalizedSales struct {
	SaleID      int     `gorm:"column:sale_id"`
	ProductName string  `gorm:"column:product_name"`
	Category    string  `gorm:"column:category"`
	StoreName   string  `gorm:"column:store_name"`
	City        string  `gorm:"column:city"`
	Amount      float64 `gorm:"column:amount"`
}

func main() {
	// Connect to the data warehouse/database
	dsn := "host=localhost user=postgres password=postgres dbname=data_warehouse port=5432 sslmode=disable"
	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{
		Logger: logger.Default.LogMode(logger.Info),
	})
	if err != nil {
		log.Printf("Failed to connect to database: %v", err)
		log.Println("Skipping actual execution...")
		return
	}

	// For demonstration, we assume tables are already populated.
	// AutoMigrate is just to ensure schema presence in this example.
	db.AutoMigrate(&DimProduct{}, &DimStore{}, &FactSales{})

	// Example: Denormalizer Engine Query
	// We use LEFT JOINs to flatten the normalized schema into our wide struct.
	// This approach offloads the join heavy-lifting to the DB engine.
	var wideTable []DenormalizedSales

	err = db.Table("fact_sales").
		Select(`
			fact_sales.sale_id,
			dim_products.product_name,
			dim_products.category,
			dim_stores.store_name,
			dim_stores.city,
			fact_sales.amount
		`).
		Joins("LEFT JOIN dim_products ON dim_products.product_id = fact_sales.product_id").
		Joins("LEFT JOIN dim_stores ON dim_stores.store_id = fact_sales.store_id").
		Scan(&wideTable).Error

	if err != nil {
		log.Fatalf("Failed to execute denormalization query: %v", err)
	}

	fmt.Println("--- Denormalized Wide Table (Ready for Analytics) ---")
	for _, rec := range wideTable {
		fmt.Printf("Sale: %d | Product: %s (%s) | Store: %s (%s) | Amount: $%.2f\n",
			rec.SaleID, rec.ProductName, rec.Category, rec.StoreName, rec.City, rec.Amount)
	}

	// In a complete ETL pipeline, this wideTable slice would then be written
	// to a fast NoSQL store (e.g., Redis, DynamoDB) for serving, or written 
	// as Parquet/CSV to S3.
}

File: DataEngineering/02_Data_Modeling/projects/python/05_denormalizer_engine/main.py
import logging
from pyspark.sql import SparkSession
from pyspark.sql.functions import col

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def main():
    """
    Denormalizer Engine using Apache Spark (PySpark).
    In a real-world scenario, this handles terabytes of data by distributing 
    the join operations across a cluster (e.g., EMR, Databricks).
    """
    # Initialize Spark Session
    spark = SparkSession.builder \
        .appName("DenormalizerEngine") \
        .master("local[*]") \
        .getOrCreate()

    spark.sparkContext.setLogLevel("ERROR")

    # 1. Load Normalized Tables (Dimensions & Facts)
    # In production, these would be loaded from S3/HDFS/Snowflake (e.g., spark.read.parquet("s3://bucket/dim_product"))
    dim_product_data = [
        (1, 'Widget A', 'Widgets'),
        (2, 'Widget B', 'Widgets')
    ]
    dim_product = spark.createDataFrame(dim_product_data, schema=['product_id', 'product_name', 'category'])
    
    dim_store_data = [
        (10, 'North Branch', 'Chicago'),
        (20, 'South Branch', 'Houston')
    ]
    dim_store = spark.createDataFrame(dim_store_data, schema=['store_id', 'store_name', 'city'])
    
    fact_sales_data = [
        (100, 1, 10, 50.0),
        (101, 2, 10, 75.0),
        (102, 1, 20, 50.0)
    ]
    fact_sales = spark.createDataFrame(fact_sales_data, schema=['sale_id', 'product_id', 'store_id', 'amount'])
    
    logger.info("--- Normalized Fact Table ---")
    fact_sales.show()
    
    # 2. Denormalizer Engine: Join everything into one wide table
    # Using left outer joins ensures we keep facts even if a dimension is missing (handling late-arriving dimensions)
    denormalized_df = fact_sales \
        .join(dim_product, on='product_id', how='left') \
        .join(dim_store, on='store_id', how='left')
                                
    logger.info("--- Denormalized Wide Table (Ready for Analytics/ML) ---")
    denormalized_df.show()

    # In production, we would write this denormalized dataframe out to an analytical store (e.g., Delta Lake, Iceberg, Parquet)
    # denormalized_df.write.format("delta").mode("overwrite").save("s3://bucket/analytics_gold_layer/denormalized_sales")

    spark.stop()

if __name__ == '__main__':
    main()
